src/Nethermind/Nethermind.Merge.Plugin/Synchronization/ChainLevelHelper.cs

```diff
@@ -46,49 +46,51 @@ public class ChainLevelHelper : IChainLevelHelper
         _syncConfig = syncConfig;
         _logger = logManager.GetClassLogger();
     }
-    
+
     public BlockHeader[] GetNextHeaders(int maxCount)
     {
         long? startingPoint = GetStartingPoint();
         if (startingPoint == null)
             return null;
-        
+
         List<BlockHeader> headers = new(maxCount);
         int i = 0;
-        
+
         while (i < maxCount)
         {
             ChainLevelInfo? level = _blockTree.FindLevel(startingPoint!.Value);
-            if (level == null)
+            if (level == null || level.MainChainBlock == null)
             {
-                if (_logger.IsTrace) _logger.Trace($"ChainLevelHelper.GetNextHeaders - level {startingPoint} not found");
+                if (_logger.IsTrace)
+                    _logger.Trace($"ChainLevelHelper.GetNextHeaders - level {startingPoint} not found");
                 break;
             }
 
-            for (int j = 0; j < level.BlockInfos.Length; ++j)
+            BlockInfo blockInfo = level.MainChainBlock;
+            BlockHeader? newHeader =
+                _blockTree.FindHeader(blockInfo.BlockHash, BlockTreeLookupOptions.TotalDifficultyNotNeeded);
+
+            if (newHeader == null)
             {
-                BlockHeader? newHeader = _blockTree.FindHeader(level.BlockInfos[j].BlockHash, BlockTreeLookupOptions.TotalDifficultyNotNeeded);
-
-                if (newHeader == null)
-                {
-                    if (_logger.IsTrace) _logger.Trace($"ChainLevelHelper - header {startingPoint} not found");
-                    continue;
-                }
-                
-                newHeader.TotalDifficulty = level.BlockInfos[j].TotalDifficulty == 0 ? null : level.BlockInfos[j].TotalDifficulty;
-                if (_logger.IsTrace) _logger.Trace($"ChainLevelHelper - A new block header {newHeader.ToString(BlockHeader.Format.FullHashAndNumber)}");
-                headers.Add(newHeader);
-                ++i;
-                if (i >= maxCount)
-                    break;
+                if (_logger.IsTrace) _logger.Trace($"ChainLevelHelper - header {startingPoint} not found");
+                continue;
             }
-            
+
+            newHeader.TotalDifficulty = blockInfo.TotalDifficulty == 0 ? null : blockInfo.TotalDifficulty;
```

Comment:
is this set to null if TD = 0 so that TD will be set later when finding / suggesting new block?

Reply:
Yes, in suggestBlock. I will add comment here

Reply:
suggestBlock calls `UpdateOrCreateLevel` to update TD which will append a NEW blockInfo to the level but with different TD. When using `FindLevel` I believe it finds the first block info with matching hash (which will be blockInfo with TD = 0 as it was inserted first). We should change this behaviour of `UpdateOrCreateLevel`?

Reply:
good catch, thank you!